lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -524,6 +524,19 @@ class TraceProcessor {
       firstMeaningfulPaint = lastCandidate;
     }
 
+    // LCP comes from the latest `largestContentfulPaint::Candidate`, but it can be invalidated
+    // by a `largestContentfulPaint::Invalidate` trace. In the case that the last candidate is
+    // invalidated, the value will be null.
+    let largestContentfulPaint;
+    for (let i = frameEvents.length - 1; i >= 0; i--) {
+      const e = frameEvents[i];
+      if (e.ts <= navigationStart.ts) break;
+      if (e.name === 'largestContentfulPaint::Invalidate') break;
```

Comment:
can we add a test for this case? 

i'd like to base the test from a real trace.

Reply:
can we get a single comment above each `if` case space shuttle style? in my first read I missed that we were counting backwards and was very confused :)

Reply:
+1 to test cases for these :)

Reply:
hmmmm luckily i just made a huge corpus of new traces.

Reply:
I ran it on a bunch of traces 
(i also put a log at `if (e.name === 'largestContentfulPaint::Invalidate')`):

```
find dist/golden-lantern-traces -name '*trace.json' | xargs -I{} node -e "\
    const path = '{}'; \
    console.log(path); \
    const trace = JSON.parse(require('fs').readFileSync(path, 'utf-8')); \
    // wpt traces start with an empty event ...
    trace.traceEvents = trace.traceEvents.filter(e => Object.keys(e).length > 0); \
    const traceOfTab = require('./lighthouse-core/lib/tracehouse/trace-processor.js').computeTraceOfTab(trace); \
    console.log(traceOfTab.timings.largestContentfulPaint)";
```

and I never hit the log point.

In fact, only two of these collected traces have `largestContentfulPaint::Invalidate`.

1. `https---www-huffpost-com--mobile-unthrottled-4-trace.json` - it's not from the main frame, so we ignore it.
1. `https---www-reddit-com--mobile-wpt-2-trace.json` - after a short gap, it's followed by a `largestContentfulPaint::Candidate`.

Reply:
IMO a real trace test for this does not need to be blocking landing (create-test-trace.js is nice :)), though if we wanted to add a flag that indicates if we fell into the invalidate path so we can examine frequency in HTTPArchive it wouldn't be the worst idea :)

Reply:
would still love some of those space shuttle comments though :D